craft_grammar/models.py

```diff
@@ -64,6 +65,100 @@ def _grammar_append(cls, entry: list[Any], item: Any, info: ValidationInfo) -> N
+        if sub_type:
+            sub_type = sub_type[0]
+            if sub_type is Any:
+                sub_type = None
+
+        new_entry: list[Any] = []
+        errors: list[pydantic_core.InitErrorDetails] = []
+        for index, item in enumerate(input_value):
+            # Check if the item is a valid grammar clause
+            try:
+                if _is_grammar_clause(item):
+                    cls._grammar_append(new_entry, item, info)
+                    continue
+            except pydantic.ValidationError as exc:
+                errors.extend(
+                    pydantic_core.InitErrorDetails(
+                        type=err["type"],
+                        loc=(index, *err["loc"]),
+                        input=err["input"],
+                        ctx=err.get("ctx", {"error": err}),
+                    )
+                    for err in exc.errors()
+                )
+                break
+            except ValueError as exc:
+                errors.append(
+                    pydantic_core.InitErrorDetails(
+                        type="value_error",
+                        loc=(index,),
+                        input=item,
+                        ctx={"error": exc},
+                    ),
+                )
+                continue
+            if sub_type:
+                sub_type_adapter = pydantic.TypeAdapter(
+                    sub_type,
+                    config=pydantic.ConfigDict(coerce_numbers_to_str=True),
+                )
+                try:
+                    new_entry.append(sub_type_adapter.validate_python(item))
+                except ValidationError:
+                    pass
+                else:
+                    continue
+            if issubclass(type_, str):
+                if isinstance(item, dict):
+                    errors.append(
+                        pydantic_core.InitErrorDetails(
+                            type="value_error",
+                            loc=(index,),
+                            input=item,
+                            ctx={
+                                "error": ValueError(
+                                    f"value must be a str or valid grammar dict: {input_value!r}",
+                                ),
+                            },
+                        ),
+                    )
+                else:
+                    raise pydantic.ValidationError.from_exception_data(
+                        title=f"Grammar[{type_.__name__}]",
+                        line_errors=[
+                            pydantic_core.InitErrorDetails(
+                                type="string_type",
+                                loc=(),
+                                input=item,
+                            ),
+                        ],
+                    )
+                    break
+            else:
+                raise ValueError(  # noqa: TRY004
+                    _format_type_error(type_, input_value),
+                )
```

Comment:
"Flat is better than nested"

```suggestion
            if not issubclass(type_, str):
                raise ValueError(  # noqa: TRY004
                    _format_type_error(type_, input_value),
                )
            if not isinstance(item, dict):
                raise pydantic.ValidationError.from_exception_data(
                    title=f"Grammar[{type_.__name__}]",
                    line_errors=[
                        pydantic_core.InitErrorDetails(
                            type="string_type",
                            loc=(),
                            input=item,
                        ),
                    ],
                )            
            errors.append(
                pydantic_core.InitErrorDetails(
                    type="value_error",
                    loc=(index,),
                    input=item,
                    ctx={
                        "error": ValueError(
                            f"value must be a str or valid grammar dict: {input_value!r}",
                        ),
                    },
                ),
            )
```

Reply:
Thanks!